### PR TITLE
Batch 2 Fast-track single chunks

### PR DESCRIPTION
**Pull request by fil512**
If a step and all prior steps produced only a single chunk of work, notify the next step immediately instead of waiting for the scheduled job maintenance to advance it.

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_0/3631-fasttrack-single-chunk-jobs.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: add
+issue: 3631
+title: "If a gated Batch 2 job step and all prior steps produced only a single chunk of work, notify the next step immediately
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_0/3631-fasttrack-single-chunk-jobs.yaml:4]: remove the 2 here, clients won't understand what it means.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDataSink.java [hunk 2]
```diff
@@ -74,4 +77,11 @@ public int getWorkChunkCount() {
 		return myChunkCounter.get();
 	}
 
+	public String getOnlyChunkId() {
+		if (getWorkChunkCount() != 1) {
+			String msg = String.format("Expected this sink to have exactly one work chunk.  Job %s v%s step %s", myJobDefinitionId, myJobDefinitionVersion, myTargetStep);
+			throw new IllegalStateException(msg);
```
Comment (tadgh) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDataSink.java:83]: thought: maybe you could also log how many chunks are present. _might_ be useful?

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java [hunk 3]
```diff
@@ -64,20 +68,46 @@ void executeStep() {
...
+				// eligible for fast tracking.
+				if (myCursor.isFinalStep()) {
+					if (updateInstanceStatus(jobInstance, StatusEnum.COMPLETED)) {
+						myJobPersistence.updateInstance(jobInstance);
+					}
+				} else {
+					JobWorkNotification workNotification = new JobWorkNotification(jobInstance, myCursor.nextStep.getStepId(), ((JobDataSink<PT,IT,OT>) dataSink).getOnlyChunkId());
+					myBatchJobSender.sendWorkChannelMessage(workNotification);
+				}
+			}
```
Comment (tadgh) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:87]: Thought: this logic is getting hairy. We are 4 `if`s deep. Not sure how to really refactor this though

**Review by tadgh — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDataSink.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobDataSink.java:83]: good call

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/JobStepExecutor.java:87]: i extracted a couple of methods to improve readability.

**Pull request merged**